web/hobbitsvc.cgi.1
.TH BB-HOSTSVC.CGI 1 "Version 4.2.1: 27 Nov 2008" "Xymon"
.SH NAME
hobbitsvc.cgi \- CGI program to view Xymon status logs
.SH SYNOPSIS
.B "hobbitsvc.cgi [--hobbitd|--historical] [--history={top|bottom}]"

.SH DESCRIPTION
\fBhobbitsvc.cgi\fR
is a CGI program to present a Xymon status log in HTML 
form (ie, as a web page). It can be used both for the logs
showing the current status, and for historical logs from the
"histlogs" directory. It is normally invoked as a
CGI program, and therefore receives most of the input parameters
via the CGI QUERY_STRING environment variable.

Unless the "--historical" option is present, the current status log
is used. This assumes a QUERY_STRING environment variable of the form
.br
   HOSTSVC=hostname.servicename
.br
where "hostname" is the name of the host with commas instead of dots,
and "servicename" is the name of the service (the column name in Xymon). 
Such links are automatically generated by the
.I bbgen(1)
tool when the environment contains "BBLOGSTATUS=dynamic".

With the "--historical" option present, a historical logfile is
used. This assumes a QUERY_STRING environment variable of the form
.br
   HOST=hostname&SERVICE=servicename&TIMEBUF=timestamp
.br
where "hostname" is the name of the host with commas instead of dots,
"servicename" is the name of the service, and "timestamp" is the time
of the log. This is automatically generated by the
.I bb-hist.cgi(1)
tool.

.SH OPTIONS
.IP "--hobbitd"
Retrieve the current status log from
.I hobbitd(1)
rather than from the logfile. This is for use with the
Xymon daemon from the Xymon monitor version 4.

.IP "--historical"
Use a historical logfile instead of the current logfile.

.IP "--history={top|bottom|none}"
When showing the current logfile, provide a "HISTORY" button
at the top or the bottom of the webpage, or not at all. The default 
is to put the HISTORY button at the bottom of the page.

.IP "--env=FILENAME"
Load the environment from FILENAME before executing the CGI.

.IP "--templates=DIRECTORY"
Where to look for the HTML header- and footer-templates used
when generating the webpages. Default: $BBHOME/web/

.IP "--no-svcid"
Do not include the HTML tags to identify the hostname/service
on the generated web page. Useful is this already happens in 
the hostsvc_header template file, for instance.

.IP "--multigraphs=TEST1[,TEST2]"
This causes hobbitsvc.cgi to generate links to service graphs 
that are split up into multiple images, with at most 5 graphs
per image. This option only works in Xymon mode. If not
specified, only the "disk" status is split up this way.

.IP "--no-disable"
By default, the info-column page includes a form allowing users 
to disable and re-enable tests. If your setup uses the default
separation of administration tools into a separate, password-
protected area, then use of the disable- and enable-functions
requires access to the administration tools. If you prefer to 
do this only via the dedicated administration page, this option
will remove the disable-function from the info page.

.IP "--no-jsvalidation"
The disable-function on the info-column page by default uses
JavaScript to validate the form before submitting the input
to the Xymon server. However, some browsers cannot handle the
Javascript code correctly so the form does not work. This 
option disables the use of Javascript for form-validation,
allowing these browsers to use the disable-function.

.SH FILES
.IP "$BBHOME/web/hostsvc_header"
HTML template header

.IP "$BBHOME/web/hostsvc_footer"
HTML template footer

.SH ENVIRONMENT
.IP "NONHISTS=info,trends,graphs"
A comma-separated list of services that does not have meaningful
history, e.g. the "info" and "trends" columns. Services listed here
do not get a "History" button.

.IP "TEST2RRD=test,test"
A comma-separated list of the tests that have an RRD graph.

.SH "SEE ALSO"
xymon(7), hobbitd(1)



Changes
Changes from 4.3.0 beta 1 -> 4.3.0 beta 2 (24 Apr 2009)
=======================================================
* New "--shuffle" option for bbtest-net to run network tests
  in a random order.
* Client startup script now exports important environment variables,
  so they are actually used in systems with a traditional shell.
* hobbitlaunch no longer crashes if there are no tasks
* Client configure script includes librt for clock_gettime()
* New client support for mainframes: z/OS, z/VM, z/VSE
* Enhanced eventlog and top-changing hosts webpages.
* Revert debian package pathnames back to use "hobbit", so
  updates from 4.2.x will actually work.
* Ghostlist options in hobbitcgi.cfg had no effect because
  of typo in setting name.
* "data" messages could crash hobbitd.
* Debug output from hobbitd_channel now logs only the relevant
  data instead of the full message.
* trimhistory now informs the history module to re-open the
  "allevents" file after trimming it.
* devmon template fix
* New "rrdcachectl" utility included.
* Fixed sorting routine (affected holiday list and others)
* Fix generic crash in communications module between Xymon
  programs, where an empty response message would crash caller.

Changes from 4.2.3 -> 4.3.0 beta 1 (09 Feb 2009)
================================================
Core changes:
* New API's for loadhosts and sendmessage, in preparation for
  the full 5.0 changes.
* Always use getcurrenttime() instead of time().
* Support for defining holidays as non-working days in alerts and
  SLA calculations.
* Hosts which appear on multiple pages in the web display can
  use any page they are on in the alerting rules and elsewhere.
* Worker modules (RRD, client-data parsers etc) can operate on
  remote hosts from the hobbitd daemon, for load-sharing.
* Various bugfixes collected over time.

Network test changes:
* Merged new network tests from trunk: SOAP-over-HTTP,
  SSL minimum cipher strength
* Changed network test code to always report a validity period
  for network tests, so it it possible to run network tests less
  often than every 30 minutes (e.g. once an hour).
* Make the content-type setting in HTTP POST tests configurable.
* Make the source-address used for TCP tests configurable.
* Make the acceptable HTTP result codes configurable.
* Use and save HTTP session cookies.

Web changes
* Support generic drop-down lists in templates.
* "NOCOLUMNS" changed to work for all columns.
* New "group-sorted" definition to auto-sort hosts in a group
* Use browser tooltips for host comments
* "Compact" status allows several statuses to appear as a single
  status on the overview webpages.
* Trends page can select the time period to show. Buttons provided
  for the common selections.
* Ghost list report now lists possible candidates for a ghost, 
  based on IP-address or unqualified hostname.

Report changes
* Number of outages as SLA parameter

Miscellaneous
* hobbitlaunch support for running tasks only on certain hosts,
  and for a maximum time.
* Alert script get a unique ID for each alert.



Changes from 4.2.2 -> 4.2.3 (09 Feb 2008)
=========================================
* Time-out code changed to use clock_gettime() with CLOCK_MONOTONIC
* Bugfix for hobbitd/hobbitd_worker communication going out-of-sync
  resulting in "garbled data" being logged and worker modules 
  stopping.
* NCV module now works with negative numbers.
* Several bugfixes in DNS lookup code - could lead to crashes when
  performing DNS tests.
* Switch to C-ARES 1.6.0 - drop support for older versions.
* Run more TCP tests in parallel by not waiting for very slow
  connections to complete before starting new ones.
* Added "hostlist" web utility for spreadsheet-reporting of the
  hosts in Hobbit.


Changes from 4.2.0 -> 4.2.2 (01 Dec 2008)
=========================================

Changelog extracted from Subversion.


------------------------------------------------------------------------
r5935 | storner | 2008-11-26 12:57:11 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   A /branches/4.2.0 (from /trunk:5040)

Create branch for maintenance of the 4.2.x release.


------------------------------------------------------------------------
r5936 | storner | 2008-11-26 13:13:50 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/common/bb-hosts.5

The ability to set a DOWNTIME for an individual test was mentioned in the Changes file, but not documented in the bb-hosts man-page.


------------------------------------------------------------------------
r5937 | storner | 2008-11-26 13:14:48 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/hobbitd/hobbitd_client.c

The hobbitd_client program would crash when running in test-mode, i.e. launched with the --test option to test disk-, procs- and ports-settings. This only applies to running it in test-mode, normal operation is not affected.


------------------------------------------------------------------------
r5938 | storner | 2008-11-26 13:15:40 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/lib/headfoot.c
   M /branches/4.2.0/web/bb-datepage.c
   M /branches/4.2.0/web/bb-eventlog.c
   M /branches/4.2.0/web/bb-findhost.c
   M /branches/4.2.0/web/bb-rep.c
   M /branches/4.2.0/web/bb-snapshot.c
   M /branches/4.2.0/web/hobbit-enadis.c
   M /branches/4.2.0/web/hobbit-nkedit.c


xymon-test/web/hobbitsvc.c
/*----------------------------------------------------------------------------*/
/* Hobbit status-log viewer CGI.                                              */
/*                                                                            */
/* This CGI tool shows an HTML version of a status log.                       */
/*                                                                            */
/* Copyright (C) 2004-2009 Henrik Storner <henrik@storner.dk>                 */
/*                                                                            */
/* This program is released under the GNU General Public License (GPL),       */
/* version 2. See the file "COPYING" for details.                             */
/*                                                                            */
/*----------------------------------------------------------------------------*/

static char rcsid[] = "$Id: hobbitsvc.c 6209 2009-08-02 11:44:02Z storner $";

#include <limits.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <ctype.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

#include "libbbgen.h"
#include "version.h"
#include "hobbitsvc-info.h"
#include "hobbitsvc-trends.h"

/* Commandline params */
static enum { SRC_HOBBITD, SRC_HISTLOGS, SRC_CLIENTLOGS } source = SRC_HOBBITD;
static int wantserviceid = 1;
static char *multigraphs = ",disk,inode,qtree,quotas,snapshot,TblSpace,if_load,";
static int locatorbased = 0;

/* CGI params */
static char *hostname = NULL;
static char *service = NULL;
static char *tstamp = NULL;
static char *nkprio = NULL, *nkttgroup = NULL, *nkttextra = NULL;
static enum { FRM_STATUS, FRM_CLIENT } outform = FRM_STATUS;
static char *clienturi = NULL;
static int backsecs = 0;
static time_t fromtime = 0, endtime = 0;

static char errortxt[1000];
static char *hostdatadir = NULL;


static void errormsg(char *msg)
{
	snprintf(errortxt, sizeof(errortxt),
		 "Content-type: %s\n\n<html><head><title>Invalid request</title></head>\n<body>%s</body></html>\n", 
		 xgetenv("HTMLCONTENTTYPE"), msg);

	errortxt[sizeof(errortxt)-1] = '\0';
}

static int parse_query(void)
{
	cgidata_t *cgidata = cgi_request();
	cgidata_t *cwalk;

	cwalk = cgidata;
	while (cwalk) {
		if (strcasecmp(cwalk->name, "HOST") == 0) {
			hostname = strdup(cwalk->value);
		}
		else if (strcasecmp(cwalk->name, "SERVICE") == 0) {
			service = strdup(cwalk->value);
		}
		else if (strcasecmp(cwalk->name, "HOSTSVC") == 0) {
			/* For backwards compatibility */
			char *p = strrchr(cwalk->value, '.');
			if (p) {
				*p = '\0';
				hostname = strdup(cwalk->value);
				service = strdup(p+1);
				for (p=strchr(hostname, ','); (p); p = strchr(p, ',')) *p = '.';
			}
		}
		else if (strcasecmp(cwalk->name, "TIMEBUF") == 0) {
			/* Only for the historical logs */
			tstamp = strdup(cwalk->value);
		}
		else if (strcasecmp(cwalk->name, "CLIENT") == 0) {
			char *p;

			hostname = strdup(cwalk->value);
			p = hostname; while ((p = strchr(p, ',')) != NULL) *p = '.';
			service = strdup("");
			outform = FRM_CLIENT;
		}
		else if (strcasecmp(cwalk->name, "SECTION") == 0) {
			service = strdup(cwalk->value);
		}
		else if (strcasecmp(cwalk->name, "NKPRIO") == 0) {
			nkprio = strdup(cwalk->value);
		}
		else if (strcasecmp(cwalk->name, "NKTTGROUP") == 0) {
			nkttgroup = strdup(cwalk->value);
		}
		else if (strcasecmp(cwalk->name, "NKTTEXTRA") == 0) {
			nkttextra = strdup(cwalk->value);
		}
		else if ((strcmp(cwalk->name, "backsecs") == 0)   && cwalk->value && strlen(cwalk->value)) {
			backsecs += atoi(cwalk->value);
		}
		else if ((strcmp(cwalk->name, "backmins") == 0)   && cwalk->value && strlen(cwalk->value)) {
			backsecs += 60*atoi(cwalk->value);
		}
		else if ((strcmp(cwalk->name, "backhours") == 0)   && cwalk->value && strlen(cwalk->value)) {
			backsecs += 60*60*atoi(cwalk->value);
		}
		else if ((strcmp(cwalk->name, "backdays") == 0)   && cwalk->value && strlen(cwalk->value)) {
			backsecs += 24*60*60*atoi(cwalk->value);
		}
		else if ((strcmp(cwalk->name, "FROMTIME") == 0)   && cwalk->value && strlen(cwalk->value)) {
			fromtime = eventreport_time(cwalk->value);
		}
		else if ((strcmp(cwalk->name, "TOTIME") == 0)   && cwalk->value && strlen(cwalk->value)) {
			endtime = eventreport_time(cwalk->value);
		}

		cwalk = cwalk->next;
	}

	if (backsecs == 0) {
		if (getenv("TRENDSECONDS")) backsecs = atoi(getenv("TRENDSECONDS"));
		else backsecs = 48*60*60;
	}

	if (!hostname || !service) {
		errormsg("Invalid request");
		return 1;
	}

	if (outform == FRM_STATUS) {
		char *p, *req;

		req = getenv("SCRIPT_NAME");
		clienturi = (char *)malloc(strlen(req) + 10 + strlen(hostname));
		strcpy(clienturi, req);
		p = strchr(clienturi, '?'); if (p) *p = '\0'; else p = clienturi + strlen(clienturi);
		sprintf(p, "?CLIENT=%s", hostname);
	}

	return 0;
}

int loadhostdata(char *hostname, char **ip, char **displayname, char **compacts)
{
	void *hinfo = NULL;

	load_hostnames(xgetenv("BBHOSTS"), NULL, get_fqdn());

	if ((hinfo = hostinfo(hostname)) == NULL) {
		errormsg("No such host");
		return 1;
	}

	*ip = bbh_item(hinfo, BBH_IP);
	*displayname = bbh_item(hinfo, BBH_DISPLAYNAME);
	if (!(*displayname)) *displayname = hostname;
	*compacts = bbh_item(hinfo, BBH_COMPACT);

	return 0;
}

int do_request(void)
{
	int color = 0;
	char timesincechange[100];
	time_t logtime = 0, acktime = 0, disabletime = 0;
	char *log = NULL, *firstline = NULL, *sender = NULL, *clientid = NULL, *flags = NULL;	/* These are free'd */
	char *restofmsg = NULL, *ackmsg = NULL, *dismsg = NULL, *acklist=NULL;	/* These are just used */
	int ishtmlformatted = 0;
	int clientavail = 0;
	char *ip, *displayname, *compacts;

	if (parse_query() != 0) return 1;

	{
		char *s;
		
		s = xgetenv("CLIENTLOGS"); 
		if (s) {
			hostdatadir = (char *)malloc(strlen(s) + strlen(hostname) + 12);
			sprintf(hostdatadir, "%s/%s", s, hostname);
		}
		else {
			s = xgetenv("BBVAR");
			hostdatadir = (char *)malloc(strlen(s) + strlen(hostname) + 12);
			sprintf(hostdatadir, "%s/hostdata/%s", s, hostname);
		}
	}

	if (outform == FRM_CLIENT) {
		if (source == SRC_HOBBITD) {
			char *hobbitdreq;
			int hobbitdresult;
			sendreturn_t *sres = newsendreturnbuf(1, NULL);

			hobbitdreq = (char *)malloc(1024 + strlen(hostname) + (service ? strlen(service) : 0));
			sprintf(hobbitdreq, "clientlog %s", hostname);
			if (service && *service) sprintf(hobbitdreq + strlen(hobbitdreq), " section=%s", service);

			hobbitdresult = sendmessage(hobbitdreq, NULL, BBTALK_TIMEOUT, sres);
			if (hobbitdresult != BB_OK) {
				char errtxt[4096];
				sprintf(errtxt, "Status not available: Req=%s, result=%d\n", hobbitdreq, hobbitdresult);
				errormsg(errtxt);
				return 1;
			}
			else {
				log = getsendreturnstr(sres, 1);
			}
			freesendreturnbuf(sres);
		}
		else if (source == SRC_HISTLOGS) {
			char logfn[PATH_MAX];
			FILE *fd;

			sprintf(logfn, "%s/%s", hostdatadir, tstamp);
			fd = fopen(logfn, "r");
			if (fd) {
				struct stat st;
				int n;

				fstat(fileno(fd), &st);
				log = (char *)malloc(st.st_size + 1);
				n = fread(log, 1, st.st_size, fd);
				if (n >= 0) *(log+n) = '\0'; else *log = '\0';
				fclose(fd);
			}
		}

		restofmsg = (log ? log : strdup("<No data>\n"));
	}
	else if ((strcmp(service, xgetenv("TRENDSCOLUMN")) == 0) || (strcmp(service, xgetenv("INFOCOLUMN")) == 0)) {
		if (loadhostdata(hostname, &ip, &displayname, &compacts) != 0) return 1;
		ishtmlformatted = 1;
		sethostenv(displayname, ip, service, colorname(COL_GREEN), hostname);
		sethostenv_refresh(600);
		color = COL_GREEN;
		logtime = getcurrenttime(NULL);
		strcpy(timesincechange, "0 minutes");

		if (strcmp(service, xgetenv("TRENDSCOLUMN")) == 0) {
			if (locatorbased) {
				char *cgiurl, *qres;

				qres = locator_query(hostname, ST_RRD, &cgiurl);
				if (!qres) {
					errprintf("Cannot find RRD files for host %s\n", hostname);
				}
				else {
					/* Redirect browser to the real server */
					fprintf(stdout, "Location: %s/bb-hostsvc.sh?HOST=%s&SERVICE=%s\n\n",
						cgiurl, hostname, service);
					return 0;
				}
			}
			else {
				if (endtime == 0) endtime = getcurrenttime(NULL);

				if (fromtime == 0) {
					fromtime = endtime - backsecs;
					sethostenv_backsecs(backsecs);
				}
				else {
					sethostenv_eventtime(fromtime, endtime);
				}

				log = restofmsg = generate_trends(hostname, fromtime, endtime);
			}
		}
		else if (strcmp(service, xgetenv("INFOCOLUMN")) == 0) {
			log = restofmsg = generate_info(hostname);
		}
	}
	else if (source == SRC_HOBBITD) {
		char hobbitdreq[1024];
		int hobbitdresult;
		char *items[20];
		int icount;
		time_t logage, clntstamp;
		char *sumline, *msg, *p, *compitem, *complist;
		sendreturn_t *sres;

		if (loadhostdata(hostname, &ip, &displayname, &compacts) != 0) return 1;

		complist = NULL;
		if (compacts && *compacts) {
			compitem = strtok(compacts, ",");
			while (compitem && !complist) {
				p = strchr(compitem, '='); if (p) *p = '\0';
				if (strcmp(service, compitem) == 0) complist = p+1;
				compitem = strtok(NULL, ",");
			}
		}

		if (!complist) {
			/* FIXME : ,flapinfo,modifies after clntstamp */
			sprintf(hobbitdreq, "hobbitdlog host=%s test=%s fields=hostname,testname,color,flags,lastchange,logtime,validtime,acktime,disabletime,sender,cookie,ackmsg,dismsg,client,acklist,BBH_IP,BBH_DISPLAYNAME,clntstamp", hostname, service);
		}
		else {
			sprintf(hobbitdreq, "hobbitdboard host=^%s$ test=^(%s)$ fields=testname,color,lastchange", hostname, complist);
		}

		sres = newsendreturnbuf(1, NULL);
		hobbitdresult = sendmessage(hobbitdreq, NULL, BBTALK_TIMEOUT, sres);
		if (hobbitdresult == BB_OK) log = getsendreturnstr(sres, 1);
		freesendreturnbuf(sres);
		if ((hobbitdresult != BB_OK) || (log == NULL) || (strlen(log) == 0)) {
			errormsg("Status not available\n");
			return 1;
		}

		if (!complist) {
			sumline = log; p = strchr(log, '\n'); *p = '\0';
			msg = (p+1); p = strchr(msg, '\n');
			if (!p) {
				firstline = strdup(msg);
				restofmsg = NULL;
			}
			else { 
				*p = '\0'; 
				firstline = strdup(msg); 
				restofmsg = (p+1);
				*p = '\n'; 
			}

			memset(items, 0, sizeof(items));
			p = gettok(sumline, "|"); icount = 0;
			while (p && (icount < 20)) {
				items[icount++] = p;
				p = gettok(NULL, "|");
			}

			/*
			 * hostname,		[0]
			 * testname,		[1]
			 * color,		[2]
			 * flags,		[3]
			 * lastchange,		[4]
			 * logtime,		[5]
			 * validtime,		[6]
			 * acktime,		[7]
			 * disabletime,		[8]
			 * sender,		[9]
			 * cookie,		[10]
			 * ackmsg,		[11]
			 * dismsg,		[12]
			 * client,		[13]
			 * acklist		[14]
			 * BBH_IP		[15]
			 * BBH_DISPLAYNAME	[16]
			 * clienttstamp         [17]
			 */
			color = parse_color(items[2]);
			flags = strdup(items[3]);
			logage = getcurrenttime(NULL) - atoi(items[4]);
			timesincechange[0] = '\0'; p = timesincechange;
			if (logage > 86400) p += sprintf(p, "%d days,", (int) (logage / 86400));
			p += sprintf(p, "%d hours, %d minutes", (int) ((logage % 86400) / 3600), (int) ((logage % 3600) / 60));
			logtime = atoi(items[5]);
			if (items[7] && strlen(items[7])) acktime = atoi(items[7]);
			if (items[8] && strlen(items[8])) disabletime = atoi(items[8]);
			sender = strdup(items[9]);

			if (items[11] && strlen(items[11])) ackmsg = items[11];
			if (ackmsg) nldecode(ackmsg);

			if (items[12] && strlen(items[12])) dismsg = items[12];
			if (dismsg) nldecode(dismsg);

			if (items[13]) clientavail = (*items[13] == 'Y');

			acklist = ((items[14] && *items[14]) ? strdup(items[14]) : NULL);

			ip = (items[15] ? items[15] : "");
			displayname = ((items[16]  && *items[16]) ? items[16] : hostname);
			clntstamp = ((items[17]  && *items[17]) ? atol(items[17]) : 0);

			sethostenv(displayname, ip, service, colorname(COL_GREEN), hostname);
			sethostenv_refresh(60);
		}
		else {
			/* Compressed status display */
			strbuffer_t *cmsg;
			char *row, *p_row, *p_fld;
			char *nonhistenv;
			char l[2048];

			color = COL_GREEN;

			cmsg = newstrbuffer(0);
			addtobuffer(cmsg, "<table width=\"80%\" summary=\"Compacted Status Info\">\n");

			row = strtok_r(log, "\n", &p_row);
			while (row) {
				/* testname,color,lastchange */
				char *testname, *itmcolor, *chgs;
				time_t lastchange;
				int icolor;

				testname = strtok_r(row, "|", &p_fld);
				itmcolor = strtok_r(NULL, "|", &p_fld);
				chgs = strtok_r(NULL, "|", &p_fld);
				lastchange = atoi(chgs);

				icolor = parse_color(itmcolor);
				if (icolor > color) color = icolor;

				sprintf(l, "<tr><td align=left>&%s&nbsp;<a href=\"%s\">%s</a></td></tr>\n", 
					itmcolor, hostsvcurl(hostname, testname, 1), testname);
				addtobuffer(cmsg, l);
				row = strtok_r(NULL, "\n", &p_row);
			}

			addtobuffer(cmsg, "</table>\n");
			ishtmlformatted = 1;

			sethostenv(displayname, ip, service, colorname(color), hostname);
			sethostenv_refresh(60);
			logtime = getcurrenttime(NULL);
			strcpy(timesincechange, "0 minutes");

			log = restofmsg = grabstrbuffer(cmsg);

			sprintf(l, "%s Compressed status display\n", colorname(color));
			firstline = strdup(l);

			nonhistenv = (char *)malloc(10 + strlen(service));
			sprintf(nonhistenv, "NONHISTS=%s", service);
			putenv(nonhistenv);
		}
	}
	else if (source == SRC_HISTLOGS) {
		char logfn[PATH_MAX];
		struct stat st;
		int fd;
		/*
		 * Some clients (Unix disk reports) dont have a newline before the
		 * "Status unchanged in ..." text. Most do, but at least Solaris and
		 * AIX do not. So just look for the text, not the newline.
		 */
		char *statusunchangedtext = "Status unchanged in ";
		char *receivedfromtext = "Message received from ";
		char *clientidtext = "Client data ID ";
		char *p, *unchangedstr, *receivedfromstr, *clientidstr, *hostnamedash;
		int n;

		if (!tstamp) { errormsg("Invalid request"); return 1; }

		if (loadhostdata(hostname, &ip, &displayname, &compacts) != 0) return 1;
		hostnamedash = strdup(hostname);
		p = hostnamedash; while ((p = strchr(p, '.')) != NULL) *p = '_';
		p = hostnamedash; while ((p = strchr(p, ',')) != NULL) *p = '_';
		sprintf(logfn, "%s/%s/%s/%s", xgetenv("BBHISTLOGS"), hostnamedash, service, tstamp);
		xfree(hostnamedash);
		p = tstamp; while ((p = strchr(p, '_')) != NULL) *p = ' ';
		sethostenv_histlog(tstamp);

		if (stat(logfn, &st) == -1) {
			errormsg("Historical status log not available\n");
			return 1;
		}

		fd = open(logfn, O_RDONLY);
		if (fd < 0) {
			errormsg("Unable to access historical logfile\n");
			return 1;
		}
		log = (char *)malloc(st.st_size+1);
		n = read(fd, log, st.st_size);
		if (n >= 0) *(log+n) = '\0'; else *log = '\0';
		close(fd);

		p = strchr(log, '\n'); 
		if (!p) {
			firstline = strdup(log);
			restofmsg = NULL;
		}
		else { 
			*p = '\0'; 
			firstline = strdup(log); 
			restofmsg = (p+1);
			*p = '\n'; 
		}


		color = parse_color(log);

		p = strstr(log, "<!-- [flags:"); 
		if (p) {
			p += strlen("<!-- [flags:");
			n = strspn(p, "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ");
			flags = (char *)malloc(n+1);
			strncpy(flags, p, n);
			*(flags + n) = '\0';
		}

		timesincechange[0] = '\0';

		p = clientidstr = strstr(restofmsg, clientidtext);
		if (p) {
			p += strlen(clientidtext);
			n = strspn(p, "0123456789");
			clientid = (char *)malloc(n+1);
			strncpy(clientid, p, n);
			*(clientid+n) = '\0';
		}

		p = unchangedstr = strstr(restofmsg, statusunchangedtext);
		if (p) {
			p += strlen(statusunchangedtext);
			n = strcspn(p, "\n"); if (n >= sizeof(timesincechange)) n = sizeof(timesincechange);
			strncpy(timesincechange, p, n);
			timesincechange[n] = '\0';
		}

		p = receivedfromstr = strstr(restofmsg, receivedfromtext); 
		if (p) {
			p += strlen(receivedfromtext);
			n = strspn(p, "0123456789.");
			sender = (char *)malloc(n+1);
			strncpy(sender, p, n);
			*(sender+n) = '\0';
		}

		/* Kill the "Status unchanged ..." and "Message received ..." lines */
		if (unchangedstr) *unchangedstr = '\0';
		if (receivedfromstr) *receivedfromstr = '\0';
	}

	if (outform == FRM_CLIENT) {
		fprintf(stdout, "Content-type: text/plain\n\n");
		fprintf(stdout, "%s", restofmsg);
	}
	else {
		if (clientid && (source == SRC_HISTLOGS)) {
			if (locatorbased) {
				char *cgiurl, *qres;

				qres = locator_query(hostname, ST_HOSTDATA, &cgiurl);
				if (!qres) {
					errprintf("Cannot find hostdata files for host %s\n", hostname);
				}
				else {
					if (clienturi) xfree(clienturi);
					clienturi = (char *)malloc(strlen(cgiurl) + 40 + strlen(hostname) + strlen(clientid));
					sprintf(clienturi, "%s/bb-hostsvc.sh?CLIENT=%s&amp;TIMEBUF=%s", 
						cgiurl, hostname, clientid);
				}
			}
			else {
				char logfn[PATH_MAX];
				struct stat st;

				sprintf(logfn, "%s/%s", hostdatadir, clientid);
				clientavail = (stat(logfn, &st) == 0);

				if (clientavail) {
					clienturi = (char *)realloc(clienturi, strlen(clienturi) + 20 + strlen(clientid));
					sprintf(clienturi + strlen(clienturi), "&amp;TIMEBUF=%s", clientid);
				}
			}
		}

		fprintf(stdout, "Content-type: %s\n\n", xgetenv("HTMLCONTENTTYPE"));
		generate_html_log(hostname, 
			  displayname,
			  service, 
			  ip,
		          color,
			  (sender ? sender : "Hobbit"), 
			  (flags ? flags : ""),
		          logtime, timesincechange, 
		          (firstline ? firstline : ""), 
			  (restofmsg ? restofmsg : ""), 
			  acktime, ackmsg, acklist,
			  disabletime, dismsg,
		          (source == SRC_HISTLOGS), 
			  wantserviceid, 
			  ishtmlformatted,
			  locatorbased,
			  multigraphs, (clientavail ? clienturi : NULL),
			  nkprio, nkttgroup, nkttextra,
			  backsecs,
			  stdout);
	}

	/* Cleanup CGI params */
	if (hostname) xfree(hostname);
	if (service) xfree(service);
	if (tstamp) xfree(tstamp);

	/* Cleanup main vars */
	if (clientid) xfree(clientid);
	if (sender) xfree(sender);
	if (flags) xfree(flags);
	if (firstline) xfree(firstline);
	if (log) xfree(log);

	return 0;
}


int main(int argc, char *argv[])
{
	int argi;
	char *envarea = NULL;

	for (argi = 1; (argi < argc); argi++) {
		if (strcmp(argv[argi], "--historical") == 0) {
			source = SRC_HISTLOGS;
		}
		else if (strcmp(argv[argi], "--hobbitd") == 0) {
			source = SRC_HOBBITD;
		}
		else if (strncmp(argv[argi], "--history=", 10) == 0) {
			char *val = strchr(argv[argi], '=')+1;

			if (strcmp(val, "none") == 0)
				histlocation = HIST_NONE;
			else if (strcmp(val, "top") == 0)
				histlocation = HIST_TOP;
			else if (strcmp(val, "bottom") == 0)
				histlocation = HIST_BOTTOM;
		}
		else if (argnmatch(argv[argi], "--env=")) {
			char *p = strchr(argv[argi], '=');
			loadenv(p+1, envarea);
		}
		else if (argnmatch(argv[argi], "--area=")) {
			char *p = strchr(argv[argi], '=');
			envarea = strdup(p+1);
		}
		else if (strcmp(argv[argi], "--no-svcid") == 0) {
			wantserviceid = 0;
		}
		else if (argnmatch(argv[argi], "--templates=")) {
			char *p = strchr(argv[argi], '=');
			sethostenv_template(p+1);
		}
		else if (argnmatch(argv[argi], "--multigraphs=")) {
			char *p = strchr(argv[argi], '=');
			multigraphs = (char *)malloc(strlen(p+1) + 3);
			sprintf(multigraphs, ",%s,", p+1);
		}
		else if (strcmp(argv[argi], "--no-disable") == 0) {
			showenadis = 0;
		}
		else if (strcmp(argv[argi], "--no-jsvalidation") == 0) {
			usejsvalidation = 0;
		}
		else if (strcmp(argv[argi], "--old-nk-config") == 0) {
			newnkconfig = 0;
		}
		else if (strcmp(argv[argi], "--debug") == 0) {
			debug = 1;
		}
		else if (argnmatch(argv[argi], "--locator=")) {
			char *p = strchr(argv[argi], '=');
			locator_init(p+1);
			locatorbased = 1;
		}
	}

	redirect_cgilog("hobbitsvc");

	*errortxt = '\0';
	hostname = service = tstamp = NULL;
	if (do_request() != 0) {
		fprintf(stdout, "%s", errortxt);
		return 1;
	}

	return 0;
}

